src/openapi/authProvider.ts

```diff
@@ -0,0 +1,419 @@
+ * @class
+ * @alias saxo.openapi.TransportAuth
+ * @param {Object} [options] - Options for auth and for the core transport. See Transport.
+ * @param {string} [options.language] - The language sent as a header if not overridden.
+ * @param {boolean} [options.defaultCache=true] - Sets the default caching behaviour if not overridden on a call.
+ * @param {string} [options.tokenRefreshUrl] - The url for refreshing the token.
+ * @param {Object.<string, string>} [options.tokenRefreshHeaders] - Any headers to be included in the token refresh call.
+ *          It should be the same format as other calls.
+ *          The object key names are the header names and the values the header values.
+ * @param {string} [options.tokenRefreshCredentials="include"] - The credentials option passed to fetch.
+ * @param {string} [options.tokenRefreshMethod="POST"] - The http method for refreshing the token e.g. "POST" or "GET" etc.
+ * @param {number} [options.tokenRefreshMarginMs=0] - The number of ms that the refresh call should be done early.
+ *          e.g. if this is 1000 it will be refreshed 1000 ms before the token expires.
+ * @param {string} [options.tokenRefreshPropertyNameToken="token"] - The property name of the token after doing a refresh.
+ * @param {string} [options.tokenRefreshPropertyNameExpires="expiry"] - The property name of the relative expiry after doing a refresh.
+ * @param {string} [options.token] - The token to use for authentication.
+ * @param {string|number} [options.expiry] - The expiry of that token, assumed to be absolute.
+ * @param {number} [options.retryDelayMs] - The delay before retrying auth
+ * @param {number} [options.maxRetryCount] - The maximum number of times to retry the auth url
+ */
+
+class AuthProvider {
+    expiry: number | string = 0;
+    token: string | null = null;
+    tokenRefreshUrl?: string
+    tokenRefreshHeaders?: Record<string, string> = {}
+    tokenRefreshCredentials = DEFAULT_TOKEN_REFRESH_CREDENTIALS
+    tokenRefreshMethod = DEFAULT_TOKEN_REFRESH_METHOD
+    tokenRefreshPropertyNameToken = DEFAULT_TOKEN_REFRESH_PROPERTY_NAME_TOKEN
+    tokenRefreshPropertyNameExpires = DEFAULT_TOKEN_REFRESH_PROPERTY_NAME_EXPIRES
+    tokenRefreshMarginMs = DEFAULT_TOKEN_REFRESH_MARGIN_MS
+    retryDelayMs = DEFAULT_RETRY_DELAY_MS
+    maxRetryCount = DEFAULT_MAX_RETRY_COUNT
+    state = 1
+    retries = 0
+    tokenRefreshTimerFireTime = 0
+    tokenRefreshTimer: any = 0
+    lastTokenFetchTime = 0
+    EVENT_TOKEN_REFRESH = 'tokenRefresh';
+    EVENT_TOKEN_RECEIVED = 'tokenReceived';
+    EVENT_TOKEN_REFRESH_FAILED = 'tokenRefreshFailed';
+    trigger: any
+
+    constructor(options?: Options) {
+        emitter.mixinTo(this);
+
+        if (!options?.token && !options?.tokenRefreshUrl) {
+            throw new Error('No token supplied and no way to get it');
+        }
+        let expiry = options?.expiry || 0;
+
+        if (options.token) {
+            this.token = addBearer(options.token);

```

Comment:
Nit: Looks like we could get rid of the surrounding `if`  if we use `addBearer(options.token || null)`. Similarly to expiry above. There's no need to set defaults then in lines 93/94.